conversion/conversion.go

```diff
@@ -200,27 +250,82 @@ func snapshotMigrationHandler(sourceProfile profiles.SourceProfile, config write
 }
 
 func dataFromDatabase(ctx context.Context, sourceProfile profiles.SourceProfile, targetProfile profiles.TargetProfile, config writer.BatchWriterConfig, conv *internal.Conv, client *sp.Client) (*writer.BatchWriter, error) {
-	infoSchema, err := GetInfoSchema(sourceProfile, targetProfile)
-	if err != nil {
-		return nil, err
-	}
-	var streamInfo map[string]interface{}
-	if sourceProfile.Conn.Streaming {
-		streamInfo, err = infoSchema.StartChangeDataCapture(ctx, conv)
-		if err != nil {
-			return nil, err
+	//handle migrating data for sharded migrations differently
+	//sharded migrations are identified via the config= flag, if that flag is not present
+	//carry on with the existing code path in the else block
+	switch sourceProfile.Ty {
+	case profiles.SourceProfileTypeConfig:
+		////There are three cases to cover here, bulk migrations and sharded migrations (and later DMS)
+		//We provide an if-else based handling for each within the sharded code branch
+		//This will be determined via the configType, which can be "bulk", "streaming" or "dms"
+		if sourceProfile.Config.ConfigType == "bulk" {
+			var bw *writer.BatchWriter
+
+			//Migrate the data from the data shards, the schema shard needs to be specified here again.
+			for _, dataShard := range sourceProfile.Config.ShardConfigurationBulk.DataShards {
+				//Create a connection profile object for it
+				fmt.Printf("Migrating shard: %v\n", dataShard.DbName)
+				infoSchema, err := getInfoSchemaForShard(dataShard, sourceProfile.Driver, targetProfile)
+				if err != nil {
+					return nil, err
+				}
+				//finally perform a snapshot migration for the shard
+				bw = performSnapshotMigration(config, conv, client, infoSchema)
+			}
+			//finally, once all shard migrations are complete, return the batch writer object
+			return bw, nil
+		} else if sourceProfile.Config.ConfigType == "dataflow" {
+			//STEP - 1 - Create batch for each physical shard
+			asyncProcessShards := func(p *profiles.DataShard, mutex *sync.Mutex) common.TaskResult[*profiles.DataShard] {
+				//create streaming cfg from the config source type.
+				streamingCfg := streaming.CreateStreamingConfig(*p)
+				//update the cfg with the HB defaults
+				err := streaming.VerifyAndUpdateCfg(&streamingCfg, targetProfile.Conn.Sp.Dbname)
+				if err != nil {
+					return common.TaskResult[*profiles.DataShard]{Result: p, Err: err}
```

Comment:
The error should contain the shard against which the error has occurred for better reporting.